pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go

```diff
@@ -403,26 +403,86 @@ func (m panicNodeMutator) Generate(
+			// causes the step to run concurrently with that step, so we filter out any concurrent steps.
+			Filter(func(s *singleStep) bool {
+				return s.context.System.Stage >= InitUpgradeStage && !idx.IsConcurrent(s)
+			})
 
-	for range numPanics {
-		nodeList := planner.currentContext.System.Descriptor.Nodes
 		targetNode := nodeList.SeededRandNode(rng)
-		addRandomly := possiblePointsInTime.
-			RandomStep(rng).
-			InsertBefore(panicNodeStep{planner.currentContext.System.Descriptor.Nodes[0], targetNode, planner.rt})
+		stepToPanic := possiblePointsInTime.RandomStep(rng)
+		hasInvalidConcurrentStep := false
+		concurrentStepMarked := false
+		var concurrentStep *singleStep
+
+		isIncompatibleStep := func(s *singleStep) bool {
+			// Restarting the system on a different node while our panicked node is still dead can
+			// cause the cluster to lose quorum, and we cannot run restartWithNewBinaryStep on the
+			// panicked node as it is already dead, so we avoid any system restarts.
+			_, restart := s.impl.(restartWithNewBinaryStep)
+			// Waiting for stable cluster version targets every node in
+			// the cluster, so a node cannot be dead during this step.
+			_, waitForStable := s.impl.(waitForStableClusterVersionStep)
+			// Many hook steps do not support running with a dead node,
+			// so we avoid inserting after a hook step.
+			_, runHook := s.impl.(runHookStep)
+
+			if idx.IsConcurrent(s) {
+				if !concurrentStepMarked {
+					concurrentStepMarked = true
+					concurrentStep = s
+				}
+				hasInvalidConcurrentStep = true
+			} else {
+				hasInvalidConcurrentStep = false
+				concurrentStepMarked = false
+			}
 
-		mutations = append(mutations, addRandomly...)
+			return restart || waitForStable || runHook
+		}
+
+		// The node should be restarted after the panic, but before any steps that are
+		// incompatible with an unavailable node, so we find the first incompatible step
+		// after the panic step and randomly insert the restart step before it.
+		_, validStartStep := upgrade.CutAfter(func(s *singleStep) bool {
+			return s == stepToPanic[0]
+		})
+		validEndStep, _, cutStep := validStartStep.Cut(func(s *singleStep) bool {
+			return isIncompatibleStep(s)
+		})
+
+		// Inserting before a concurrent step will cause the step to run concurrently with that step,
+		// so we remove the concurrent steps from the list of possible insertions if they contain
+		// any invalid steps.
+		if hasInvalidConcurrentStep {
+			validEndStep, _, _ = validEndStep.Cut(func(s *singleStep) bool {
+				return s == concurrentStep
+			})
+		}
+
+		restartDesc := fmt.Sprintf("restarting node %d after panic", targetNode[0])
+
+		addPanicStep := stepToPanic.
+			InsertBefore(panicNodeStep{planner.currentContext.System.Descriptor.Nodes[0], targetNode})
+		var addRestartStep []mutation
+		// If validEndStep is nil, it means that there are no steps after the panic step (not including the step we insert
```

Comment:
Not sure what `(not including the step we insert before the panic)` is referring to here. I think the comment makes sense without it anyway?